### PR TITLE
Fix ListBox items accessible objects memory leak 

### LINKED ISSUE #7197: ListBox items' accessible objects are leaking in memory
### .NET version

.NET 7.0 Preview 4 

### Did it work in .NET Framework?

No

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

No

### Issue description

`ListBox` items' accessible objects are left in memory after the `ListBox` control instance with its items was inspected with accessibility tool (like Accessibility Insights or Inspect) and the form with `ListBox` is closed, garbage is collected, but accessibility tool is still opened. Same applies for case when `ListBox` item is removed from `ListBox`.

Memory snapshot in VS:
Screenshot 2022-05-18 174214

### Steps to reproduce

1. Create a parent form with two buttons: one will be opening a child form with the `ListBox` control and the other will collect the garbage.
2. Press the second button to open the child form with `ListBox`.
3. Start the Accessibility Insights tools.
4. Run automated checks for the `ListBox` control using AI.
5. Close the child form.
6. Clear memory with the first button of the parent form.
7. DON'T close the AI tool.
8. Look for `ListBoxItemAccessibleObject` accessible objects left in memory (using WinDbg, or dotMemory, or another suitable tool).

Sample app: [ListBoxSample.zip](https://github.com/dotnet/winforms/files/8731986/ListBoxSample.zip)

### PR DESCRIPTION
**Pull request by dmitrii-drobotov**
Fixes #7197.

## Proposed changes

- Add call to `UiaDisconnectProvider` for `ListBoxItemAccessibleObject`s for cases when:
    - `ListBox` is destroyed (now it only releases provider for `ListBoxAccessibleObject`)
    - All items are cleared from `ListBox`
    - An item is removed from `ListBox`
- Add unit tests for above scenarios

## Customer Impact

- `ListBoxItemAccessibleObject`s memory will stop leaking 

## Regression? 

- No

## Risk

- Minimal

### Before

After executing steps from issue description for ListBox with 18 items:

Screenshot 2022-06-09 184103

### After

Screenshot 2022-06-09 180310

After `ListBox.Items.Clear()`:

Screenshot 2022-06-09 183627

One `ListBoxItemAccessibleObject` instance may be held by `ListBox._focusedItem` field, which doesn't create a memory leak, because the AO was disconnected from UIA provider and will be garbage-collected normally once form is closed.

After  `ListBox.Items.RemoveAt(0)` 5 times:

Screenshot 2022-06-09 183839

## Test methodology 

- Unit testing
- Manual testing with WinDbg to check memory leak, Narrator to check for regressions similar to #6230, Inspect/AI to check that removed items are inaccessible there
 

## Test environment(s)

* Windows 11 (Version 10.0.22000)
* .NET 7.0.0-preview.5.22272.3

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7284)

### CODE REVIEW

**Review by RussKie — COMMENTED** [review]
👍

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.AccessibleObject.cs [hunk 1]
```diff
@@ -177,9 +177,41 @@ internal override bool IsPatternSupported(UiaCore.UIA patternId)
 
             internal void ResetListItemAccessibleObjects()
             {
+                if (OsVersion.IsWindows8OrGreater)
+                { 
+                    foreach (ListBoxItemAccessibleObject itemAccessibleObject in _itemAccessibleObjects.Values)
+                    {
+                        HRESULT result = UiaCore.UiaDisconnectProvider(itemAccessibleObject);
+                        Debug.Assert(result == HRESULT.S_OK);
```
[the remark this answers is not in the document]
Comment (dmitrii-drobotov) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListBox.AccessibleObject.cs:185]: This extension method would miss `HRESULT.S_FALSE` which has value of 1. See https://github.com/dotnet/winforms/pull/6249#discussion_r870798274

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs [hunk 2]
```diff
@@ -2091,6 +2099,20 @@ protected override void RefreshItem(int index)
             Items.SetItemInternal(index, Items[index]);
         }
 
+        internal override void ReleaseUiaProvider(IntPtr handle)
+        {
+            base.ReleaseUiaProvider(handle);
+            ClearListItemAccessibleObjects();
+        }
+
+        private void RemoveListItemAccessibleObject(int index)
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:2108]: Consider using `At` in the name end - `RemoveListItemAccessibleObjectAt` like we use for `RemoveAt`

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.AccessibleObject.cs [hunk 3]
```diff
@@ -177,9 +177,41 @@ internal override bool IsPatternSupported(UiaCore.UIA patternId)
...
+                    }
+                }
+
                 _itemAccessibleObjects.Clear();
             }
 
+            internal void RemoveListItemAccessibleObject(int index)
+            {
+                if (index >= _owningListBox.Items.InnerArray.Entries.Count)
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.AccessibleObject.cs:194]: Extract entries:
```suggestion
            {
                ItemArray.Entry[] entries = _owningListBox.Items.InnerArray.Entries;
                if (index >= entries.Count)
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.AccessibleObject.cs [hunk 4]
```diff
@@ -177,9 +177,41 @@ internal override bool IsPatternSupported(UiaCore.UIA patternId)
...
+                    {
+                        HRESULT result = UiaCore.UiaDisconnectProvider(itemAccessibleObject);
+                        Debug.Assert(result == HRESULT.S_OK);
+                    }
+                }
+
                 _itemAccessibleObjects.Clear();
             }
 
+            internal void RemoveListItemAccessibleObject(int index)
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.AccessibleObject.cs:192]: Consider as well
```suggestion
            internal void RemoveListItemAccessibleObjectAt(int index)
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs [hunk 5]
```diff
@@ -2091,6 +2099,20 @@ protected override void RefreshItem(int index)
             Items.SetItemInternal(index, Items[index]);
         }
 
+        internal override void ReleaseUiaProvider(IntPtr handle)
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:2102]: Please check the RecreateControl case. Are the objects correct after recreating the ListBox control?
For example, open PropertyGrid, use some objects as SelectedObject, and check AccessibleRole property, expand its items and make sure that they are correct in inspect after several expand-collapse actions.
After that also check memory leaks of ListBox after disposing of the PropertyGrid.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListBoxAccessibleObjectTests.cs [hunk 6]
```diff
@@ -102,5 +93,72 @@ public void ListBoxItemAccessibleObject_GetPropertyValue_ValueValuePropertyId_Re
...
+                "g"
+            });
+
+            return listBox;
+        }
+
+        private ListBoxAccessibleObject InitListBoxItemsAccessibleObjects(ListBox listBox)
+        {
+            var accessibilityObject = (ListBoxAccessibleObject)listBox.AccessibilityObject;
+            var childCount = accessibilityObject.GetChildCount();
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListBoxAccessibleObjectTests.cs:152]: Explicit `var`s

**Review by vladimir-krestov — COMMENTED** [review]
Looks good! Check recreating case

**Comment by vladimir-krestov** [comment]
PS. Tested the fix for a `PropertyGrid` case (`GridViewListBoxAccessibleObject`). The fix works correctly, 0 ListBox objects are leaking. ✔️

File: src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs (hunk 5 above)
Comment (dmitrii-drobotov) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:2102]: I checked it and can confirm that there are no issues after recreation. `ReleaseUiaProvider` indeed gets called and items become unavailable in Inspect (same as if form was closed) but once you click Refresh, new items will be in place. In AI there is an auto-refresh, so new list simply replaces the old one. Narrator works fine as well. It also doesn't cause memory leak.

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
@msftbot merge if @Tanya-Solyanik approves

**Review by vladimir-krestov — APPROVED** [review]

**Comment by dmitrii-drobotov** [comment]
@vladimir-krestov For `CheckedListBox` it also fixes the leak.
Before fix:
image
After fix:
Screenshot 2022-06-27 110051

**Review by Tanya-Solyanik — APPROVED** [review]
Looks good!

**Pull request merged**